go/libraries/doltcore/sqle/dolt_index.go

```diff
@@ -70,63 +68,84 @@ func (di *doltIndex) NewLookup(ctx *sql.Context, ranges ...sql.Range) (sql.Index
-			var keys2 []interface{} // used only when both bounds are set, thus will always represent the upperbound
-
-			for _, rangeColumn := range rang {
-				if len(rangeTypes) != len(rangeColumn) {
-					return nil, nil //TODO: support indexes having different range counts
-				}
-				rangeColumnExpr := rangeColumn[i]
-				if rangeColumnExpr.Type() != rangeType {
-					return nil, nil //TODO: support mixing range types
-				}
+		rangeCheck := make(nomsRangeCheck, len(rang))
+		for i, rangeColumnExpr := range rang {
+			// An empty column expression will mean that no values for this column can be matched, so we can discard the
+			// entire range.
+			if ok, err := rangeColumnExpr.IsEmpty(); err != nil {
+				return nil, err
+			} else if ok {
+				continue RangeLoop
+			}
 
-				hasLower := rangeColumnExpr.HasLowerBound()
-				hasUpper := rangeColumnExpr.HasUpperBound()
-				if hasLower && hasUpper {
-					keys1 = append(keys1, sql.GetRangeCutKey(rangeColumnExpr.LowerBound))
-					keys2 = append(keys2, sql.GetRangeCutKey(rangeColumnExpr.UpperBound))
-				} else if hasLower && !hasUpper {
-					keys1 = append(keys1, sql.GetRangeCutKey(rangeColumnExpr.LowerBound))
-				} else if !hasLower && hasUpper {
-					keys1 = append(keys1, sql.GetRangeCutKey(rangeColumnExpr.UpperBound))
+			cb := columnBounds{}
+			promotedType := di.cols[i].TypeInfo.Promote()
+			if rangeColumnExpr.HasLowerBound() {
+				key := sql.GetRangeCutKey(rangeColumnExpr.LowerBound)
+				val, err := promotedType.ConvertValueToNomsValue(ctx, di.table.ValueReadWriter(), key)
+				if err != nil {
+					return nil, err
 				}
+				cb.lowerbound.equals = rangeColumnExpr.LowerBound.TypeAsLowerBound() == sql.Closed
+				cb.lowerbound.infinity = false
+				cb.lowerbound.val = val
+			} else {
+				cb.lowerbound.equals = false
+				cb.lowerbound.infinity = true
+				cb.lowerbound.val = nil
 			}
-
-			lookupRange, err := idx.sqlRangeToLookupRange(ctx, rangeType, keys1, keys2)
-			if err != nil {
-				return nil, err
+			if rangeColumnExpr.HasUpperBound() {
+				key := sql.GetRangeCutKey(rangeColumnExpr.UpperBound)
+				val, err := promotedType.ConvertValueToNomsValue(ctx, di.table.ValueReadWriter(), key)
+				if err != nil {
+					return nil, err
+				}
+				cb.upperbound.equals = rangeColumnExpr.UpperBound.TypeAsUpperBound() == sql.Closed
+				cb.upperbound.infinity = false
+				cb.upperbound.val = val
+			} else {
+				cb.upperbound.equals = false
+				cb.upperbound.infinity = true
+				cb.upperbound.val = nil
 			}
-			lookupRanges = append(lookupRanges, lookupRange)
+			rangeCheck[i] = cb
 		}
+
+		readRanges = append(readRanges, &noms.ReadRange{
+			Start:     lowerboundTuple,
+			Inclusive: inclusive,
+			Reverse:   false,
+			Check:     rangeCheck,
+		})
 	}
 
 	return &doltIndexLookup{
-		idx:       idx,
-		ranges:    lookupRanges,
+		idx:       di,
+		ranges:    readRanges,
```

Comment:
Possible for readRanges to be empty at this point? Does that just mean an empty lookup, working as intended?

Reply:
Correct, nil ranges will be an empty lookup returning no rows as intended.